Add withStateHandlers()

helixbass opened the pull request:
Fixes #90 

Adds support for `withStateHandlers()`

I'm not highly familiar with `recompact` codebase but pieced this together by looking at implementations of `withState()`, `withHandlers()`, and `recompose`'s `withStateHandlers()` implementation (as well as corresponding tests)

Comparing to `recompose`'s tests:
- Omitted `withStateHandlers initial state must be function or object or null or undefined`, which checks that mounting with `initialState` `1` throws an error (because it uses `.setState()`, which we don't here)
- Omitted `withStateHandlers should persist events passed as argument`, since we're not using a `.setState()` callback there seems to be no need to persist events
- Omitted `withStateHandlers does not rerender if state updater returns undefined` (and related following test), as well as the corresponding documentation "Returning undefined does not cause a component rerender."
  - If this behavior (of not rerendering) is desirable, it seems like perhaps it would be possible (though I'm not familiar with how `updateProps()` works) to do this by caching the last object passed to `next()` and then passing the cached object to `next()` if the handler (`handler(state, props)(...args)`) returns undefined?

Included documentation in `withStateHandlers` module but didn't run eg `yarn run build`, please let me know if there are any scripts like that which I should run for the purposes of this PR, thanks

matthieuprat commented:
Thank you! I'll look into it this weekend.

neoziro commented:
Thanks @helixbass and @matthieuprat

matthieuprat commented on src/withStateHandlers.js:
```diff
@@ -0,0 +1,70 @@
+    let state
+
+    const handlers = mapValues(stateUpdaters, handler => (...args) => {
+      state = { ...state, ...handler(state, props)(...args) }
```
Recompose calls `persist` on the first argument if it is a SyntheticEvent (acdlite/recompose#476). I think we want to go the same way. Would you be willing to address this?

helixbass commented on src/withStateHandlers.js:
```diff
@@ -0,0 +1,77 @@
+    const handlers = mapValues(
+      stateUpdaters,
+      handler => (mayBeEvent, ...args) => {
+        if (mayBeEvent && typeof mayBeEvent.persist === 'function') {
```
@matthieuprat sure, updated to call `persist()` on `SyntheticEvent` like Recompose and added corresponding test (from Recompose), the only reason I didn't include this originally is that we don't call `setState()` (using callback) like Recompose so I didn't know if it was necessary, but without fully understanding `recompact` internals it seems safer to `persist()`

matthieuprat replied:
Oh, you're right, I misread the PR description. Indeed, we don't need to persist the event since we're always calling the handler synchronously. Sorry for the confusion.

matthieuprat commented:
> Omitted `withStateHandlers initial state must be function or object or null or undefined`, which checks that mounting with `initialState` `1` throws an error (because it uses `.setState()`, which we don't here)

Fine by me.

> Omitted `withStateHandlers should persist events passed as argument`, since we're not using a `.setState()` callback there seems to be no need to persist events

Indeed.

> [...] If this behavior (of not rerendering) is desirable, it seems like perhaps it would be possible (though I'm not familiar with how `updateProps()` works) to do this by caching the last object passed to `next()` and then passing the cached object to `next()` if the handler (`handler(state, props)(...args)`) returns undefined?

I think this behavior is desirable. To achieve this, we would just need to not call `next` if the partial state returned by `handler(...)(...args)` is falsy.

helixbass commented:
@matthieuprat updated to not persist event and to not call `next()` when the return value of the state handler `== null` (this seems to match the behavior of `setState()` per [this Twitter conversation](https://twitter.com/dan_abramov/status/852553760358502400))

Added `sinon` as a dev dependency to be able to mimic the corresponding Recompose test for not rerendering

neoziro commented:
@matthieuprat I let you manage it, just keep me posted to release a new version of recompact

matthieuprat commented:
@helixbass

> updated to not persist event and to not call next() when the return value of the state handler == null (this seems to match the behavior of setState() per this Twitter conversation)

I came across the same Twitter thread but it seems that React merely checks that the partial state returned by the state updater is falsy. See https://github.com/facebook/react/blob/v16.2.0/packages/react-reconciler/src/ReactFiberUpdateQueue.js#L284.

> Added sinon as a dev dependency to be able to mimic the corresponding Recompose test for not rerendering

Couldn't you get the desired behavior with `jest.fn`?

matthieuprat commented:
@neoziro Sure.

helixbass commented:
@matthieuprat didn't realize you could achieve with `jest.fn()`, updated to remove `sinon`. And updated to just use falsiness rather than `== null` to suppress calling `next()`

matthieuprat commented:
Perfect, thanks!
